### PR TITLE
core: Support layered provider names in FI_PROVIDER

### PR DESCRIPTION
**Pull request by j-xiong**
The environment variable FI_PROVIDER is a comma separated list of provider
names that are allowed to be used. Previously layered provider names were
not supported and a definition such as FI_PROVIDER="verbs;ofi_rxm" would
eliminate all providers because any core provider would fail to register
due to not matching the name "verbs;ofi_rxm".

This patch extends the FI_PROVIDER usage to allow layered provider names
while remain fully backward compatible with previous implementation in all
other aspects.

Signed-off-by: Jianxin Xiong <jianxin.xiong@intel.com>

### CODE REVIEW

File: prov/mrail/src/mrail_init.c [hunk 1]
```diff
@@ -165,7 +165,7 @@ int mrail_get_core_info(uint32_t version, const char *node, const char *service,
 		FI_DBG(&mrail_prov, FI_LOG_CORE,
 		       "--- Begin fi_getinfo for rail: %zd ---\n", i);
 
-		ret = fi_getinfo(version, NULL, NULL, 0, core_hints, &info);
+		ret = fi_getinfo(version, NULL, NULL, OFI_GETINFO_INTERNAL, core_hints, &info);
```
Comment (shefty) [inline prov/mrail/src/mrail_init.c]: I think these changes should be pulled into a separate patch.

Along this line, what prevents mrail from recursively calling getinfo to itself?  The CORE_PROV_ONLY flag is used by the util providers, but that's not used by mrail.

File: src/fabric.c [hunk 2]
```diff
@@ -63,6 +63,52 @@ int ofi_init = 0;
...
+{
+	int i, len;
+
+	if (!filter || !filter->names || !filter->names[0])
+		return 1;
+
+	for (i=0; filter->names[i]; i++) {
+		len = strlen(filter->names[i]);
+		if (!strncmp(prov_name, filter->names[i], len) &&
+		    (prov_name[len] == '\0' || prov_name[len] ==';'))
```
Comment (shefty) [inline src/fabric.c]: For the post filter, I was expecting a simple string comparison match.  Why are the extra prov_name checks here?

File: src/fabric.c [hunk 3]
```diff
@@ -463,6 +509,17 @@ void fi_ini(void)
...
+
+	/* remove layering from the base filter */
+	if (prov_filter.names) {
+		int i;
+		char *p;
+		for (i=0; prov_filter.names[i]; i++) {
+			if ((p = strchr(prov_filter.names[i], ';')))
+				*p = '\0';
+		}
+	}
```
Comment (shefty) [inline src/fabric.c]: I wonder if we should move the creation of the provider filters to a new call.

File: prov/mrail/src/mrail_init.c (hunk 1 above)
Comment (j-xiong) [reply prov/mrail/src/mrail_init.c]: CORE_PROV_ONLY can't be used here because mrail can be layered over other utility providers. Recursion is not going to happen because "^ofi_mrail" is explicitly added to core_hints->fabric_attr->prov_name.

This part is included here to get mrail layering to work (otherwise the intermediate result is filtered out).

File: src/fabric.c (hunk 2 above)
Comment (j-xiong) [reply src/fabric.c]: the extra check allows extra layering. for example, if user asks for "verbs;ofi_rxm" then both "verbs;ofi_rxm" and "verbs;ofi_rxm;ofi_mrail" will match, but should match "verbs;ofi_rxm2".

File: prov/mrail/src/mrail_init.c (hunk 1 above)
Comment (shefty) [reply prov/mrail/src/mrail_init.c]: I understood why CORE_PROV flag couldn't be used.  I forgot that we have the ^mrail option to avoid recursion

File: src/fabric.c (hunk 3 above)
Comment (j-xiong) [reply src/fabric.c]: It was originally that way. I can change it back if that is more preferable.

**Comment by shefty** [comment]
I thought about this some more.  For the standard filter, would updating the generic filter code to support both ',' and ';' as separators work?  Then the post filter performs a strcmp against the environment string.

This doesn't quite give us the right behavior if the names are negated (valid options would be tossed out), so an alternative would be to check for ';' and only create the post filter in that case.

File: src/fabric.c (hunk 2 above)
Comment (shefty) [reply src/fabric.c]: I didn't even consider the case where the user could specify "verbs;ofi_rxm,verbs;ofi_rxd".  Does the change handle that?

The extra checks make sense now.  Thanks

**Comment by j-xiong** [comment]
The PR is updated. Now a single filter is used for both standard filtering and post filtering. Different "find_name" functions are defined to handle different matching requirements.

File: src/fabric.c [hunk 4]
```diff
@@ -99,13 +125,42 @@ static int ofi_is_hook_prov(const struct fi_provider *provider)
...
-		return filter->negated ? 0 : 1;
-	}
-	return 0;
+	if (ofi_find_name(filter->names, name) >= 0)
+		return filter->negated ? 1 : 0;
+
+	return filter->negated ? 0 : 1;
+}
+
+int ofi_apply_prov_init_filter(struct fi_filter *filter, const char *name)
```
Comment (shefty) [inline src/fabric.c]: Okay, I've tried to read this over and over, and I've come to the conclusion that I think we need comments above the new calls that describe how the new functions' filters work.

Comment (j-xiong) [reply src/fabric.c]: I have added comments to the two new "apply_xxx_filter" functions. Hope that helps.
